icefoganalytics/travel-authorization
# Part 4: Wire Travel Desk Request Tab

Relates to:

- Part of solving https://github.com/icefoganalytics/travel-authorization/issues/167
- [TravelDeskWireFrame.pdf](https://github.com/icefoganalytics/travel-authorization/files/14016343/TravelDeskWireFrame.pdf)
- [UserStories-TravelDesk.docx](https://github.com/icefoganalytics/travel-authorization/files/14016344/UserStories-TravelDesk.docx)
- [TravelDeskDataModel (1).xlsx](https://github.com/icefoganalytics/travel-authorization/files/14016345/TravelDeskDataModel.1.xlsx)
- https://github.com/ytgov/travel-authorization/tree/main/web/src/modules/travelDesk
- https://github.com/icefoganalytics/travel-authorization/pull/182
- https://github.com/icefoganalytics/travel-authorization/pull/183
- https://github.com/icefoganalytics/travel-authorization/pull/185

# Context

Using the existing Travel Desk forms (https://github.com/ytgov/travel-authorization/tree/main/web/src/modules/travelDesk) rewire the travel desk request to exist in the travel desk tab of the travel auth.

Travel Desk forms should be created anytime a user is traveling by air.

![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/74adc941-50cd-4219-a723-062822c4a94a)

# Implementation

- Genericized `TravelDeskHotel` edit table and children, so they can work outside of their current location. 
- Standardized the hotels table structure and updated the entity relationship diagrams.
- Made paranoid mode the default for all models, enhancing data security.
-  Alphabetize model imports.
- Add "blocked to true confirm" dialog helper, that lets the user keep deleting table entries after disabling the confirmation dialog.

## Concerns

Selecting "no" for conference meeting/hotel, still requires that you enter a hotel/conference name.
This functionality was in the original code, but doesn't quite make sense to me.
It might make sense to either update the text on change?

# Screenshots

Hotel Request edit table
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/e97bab8f-b338-4edc-bf7d-fbe3a42e613d)

Create hotel dialog
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/1c2a0aff-db17-4735-846b-75c71846352e)

Edit Hotel dialog
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/214b6bbe-d028-4c79-9030-7df87f1057d3)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Go to the "My Travel Requests" page via the top drop down nav.
5. Create a travel authorization.
6. In the details section, select at least one travel method with "aircraft" type.
7. Generate the travel estimate.
8. Submit the travel authorization to yourself as the supervisor.
9. Go to Manger View page accessible via the drop down nav.
10. Click on the new request from the "Pending Approvals" panel in the top left.
11. Scroll to the bottom and approve the request.
12. Go back to the "My Travel Requests" page, and find your new travel request in the table.
13. Click on the Submit Travel Desk Request button to go to the Request tab.
14. Add one flight in the Flight Request section.
15. Scroll down to the "Add hotel" section and click the "Add Hotel" button to open the dialog.
16. Note that only the Check-in Date is prefilled from the flight.
17. Go back to the flights section and add another flight.
18. Go back to the hotels section and open the add hotel dialog again. Note that both check-in/out are prefilled.
19. Fill in the rest of the required fields. Check that validation works as expected.
20. After creating a hotel, test out editing it via the "pencil" icon in the table.
21. Check that you can delete a hotel via the red "x" icon.

Note that the Other Transportation Request section do not work yet.
